### DISCUSSION TITLE
Start with an older fully downloaded copy of the photos library, and end with a recent partly-downloaded copy?

### DISCUSSION THREAD

**Discussion by swduncan** [RhetTbull/osxphotos#989] [Q&A]
I have a fully-downloaded copy of my photos library, and I have the current system library which began as the fully-downloaded library mentioned before, but set to optimize space on my Mac. Of course there are new photos that have added since the old copy was made. 

If I export the first fully-downloaded library and then switch the source the current library and use --update will it see the existing files from the first library as 'foreign' or part of the same library?

Thanks!

**Comment by RhetTbull** [comment]
That's an interesting scenario. Unfortunately, osxphotos doesn't support this situation as currently designed. osxphotos tracks assets in the library using a universally unique identifier (UUID) which Photos assigns to each asset upon import. However, this unique identifier is unique only to the copy of the Photos library.  If you rebuild the library or have a copy of the library from a different computer, the UUIDs won't match and thus Photos won't be able to match the exported photos to those in the 2nd library. 

There are a couple of ways osxphotos could solve this:
1. For iCloud photos (your situation) each photo has a "cloud master UUID" that maps the photo to the original in iCloud. osxphotos doesn't currently read this value but if it did, this could be used to compare the photos in the two databases to find the matching photos.
2. For all photos, including those not in the cloud, Photos computes a "fingerprint" hash that is unique to each asset and is the same across computers/libraries. osxphotos already uses this to find identical photos for the `osxphotos sync` command. This could be used to find the matching photos (though there is some ambiguity as you could have duplicate photos in the library and they'd have the same fingerprint). 

Given one or both of the above, it would be possible to add a new command to `osxphotos exportdb` that would update the export database with new UUIDs from a new library. In addition to your situation, this would be helpful for those that want to maintain an existing export after moving to a new computer (the photos, if downloaded from iCloud and not restored from a backup, will have different UUIDs). 

I'll think about this but I think it's do-able and would be useful.  I'll open a new issue to add this feature.

**Comment by RhetTbull** [comment]
@swduncan I've published release [0.57.3](https://github.com/RhetTbull/osxphotos/releases/tag/v0.57.3) which adds a new option to address your use case.  You can now use `osxphotos exportdb --migrate-photos-library` to change the Photos library associated with an export which should let you export from the old library then update from the new library, with some caveats.

To do this:

1. Update osxphotos to release 0.57.3
2. Run an `osxphotos export --update` with the original library used for your export. This step is required after upgrading to ensure that all the metadata in the export database is refreshed, even if no files are exported.
3. Run `osxphotos exportdb /path/to/export/folder-or-export-database --migrate-photos-library /path/to/new/photos/library`
4. For step 3, you can add `--verbose --dry-run` if desired to see what will be changed. osxphotos will attempt to match all photos in the export directory to the correct photos in the new library and update the UUIDs in the export database. A backup of the export database will be created with the `.YYYYMMDDHHMMSS.bak` extension.
5. Now run an export with the new library.

Caveats: 
1. Some photos will still get re-exported as the metadata or signature may not match 100% between the old library and new library. 
2. If you've exported shared photos, they will all likely get re-exported unless your `--filename` template gives them a unique name. This is because the default name for shared photos is the UUID of the photo as the original filename (e.g. `IMG_1234.jpg`) is not available for shared photos. The previously exported shared photos will thus appear to be different photos.
3. If you used `--exiftool`, it's possible these photos will need to be re-exported as the metadata signatures won't match.  The `--migrate-photos-library` command takes steps to avoid this by updating the metadata hash in the export database but this is not 100% guaranteed to work in every case.
